File: Algorithm_443_String Compression.cpp
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    int compress(vector<char>& a) {
        int i = 0, t = 0;
        for (int j = 0; j < a.size(); ++j) if (j+1 == a.size() || a[j+1] != a[j]) {
            int m = t;
            a[t++] = a[j];
            if (j >= i + 1) {
                string s = to_string(j-i+1);
                for (char c:s) a[t++] = c;
            }
            i = j + 1;
        }
        return t;
    }
};

File: Algorithm_443_String Compression.java
// 双指针，cur记录遍历原始chars的指针的当前位置
// newLen记录用于覆盖chars中原始内容的指针的位置

class Solution {
    public int compress(char[] chars) {
        int cur = 0, newLen = 0;
        while (cur < chars.length) {
            char c = chars[cur];
            int count = 1;
            while (cur < chars.length - 1 && chars[cur + 1] == c) {
                cur++;
                count++;
            }
            cur++; // move to the new letter
            chars[newLen++] = c;
            if (count > 1) { // multiple chars
                // append the count
                String countStr = String.valueOf(count);
                for (int i = 0; i < countStr.length(); i++) {
                    chars[newLen++] = countStr.charAt(i);
                }
            }
        }
        return newLen;
    }
}